drivers/saul/init_devs/auto_init_saul_pwm.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ */
+
+/*
+ * @ingroup     sys_auto_init_saul
+ * @{
+ *
+ * @file
+ * @brief       Auto initialization of PWM pins directly mapped to SAUL reg
+ *
+ * @author      Christian Amsüss <chrysn@fsfe.org>
+ *
+ * When this module is used, any PWM device assigned inside the configuration
+ * structs inside `pwm_params.h` in the @ref saul_pwm_dimmer_params_t and @ref
+ * saul_pwm_rgb_params_t is initialized at startup for 8-bit dimming at about
+ * 1kHz, and the indicated channels are exposed via SAUL.
+ *
+ * @}
+ */
+
+#include "log.h"
+#include "saul_reg.h"
+#include "saul/periph.h"
+#include "pwm_params.h"
+#include "periph/pwm.h"
+
+#define MAX(a, b) (((a) > (b)) ? (a) : (b))
+
+#if !defined(SAUL_PWM_FREQ)
+#define SAUL_PWM_FREQ SAUL_PWM_FREQ_DEFAULT
+#endif
+
+/**
+ * @brief   Define the number of configured dimmers
+ */
+#ifndef SAUL_PWM_NO_DIMMER
+#define SAUL_PWM_DIMMER_NUMOF ARRAY_SIZE(saul_pwm_dimmer_params)
+#else
+#define SAUL_PWM_DIMMER_NUMOF 0
+static const saul_pwm_dimmer_params_t saul_pwm_dimmer_params[1];
+#endif
+
+/**
+ * @brief   Define the number of configured RGB LEDs
+ */
+#ifndef SAUL_PWM_NO_RGB
+#define SAUL_PWM_RGB_NUMOF ARRAY_SIZE(saul_pwm_rgb_params)
+#else
+#define SAUL_PWM_RGB_NUMOF 0
+static const saul_pwm_rgb_params_t saul_pwm_rgb_params[1];
+#endif
+
+/**
+ * @brief   Memory for the registry RGB LED entries
+ */
+/* The static variable will be unused in the 0 case and thus not emitted. */
+static saul_reg_t saul_reg_entries_rgb[MAX(SAUL_PWM_RGB_NUMOF, 1)];
```

Comment:
GCC has no issue with zero length arrays, unless one compiles with `-pedantic`. `clang` is also not picky about them. Dropping the `MAX()` would simplify the code a bit, but I'm not sure if this is a good idea.

Reply:
Oh, I was unaware of that. I usually work in C with -pedantic settings, but if we don't insist on that here, I'm happy to make the simplifications that ensue.